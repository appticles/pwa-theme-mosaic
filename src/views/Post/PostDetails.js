import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import Helmet from 'react-helmet';
import { Container, Header, Icon, Modal, Transition } from 'semantic-ui-react';
import Moment from 'react-moment';
import PropTypes from 'prop-types';
import _ from 'lodash';
import config from 'react-global-configuration';
import { DFPSlotsProvider, AdSlot } from 'react-dfp';
import SocialMedia from './components/SocialMedia';
import { postPropType } from '../PostsCarousel/reducer';
import CategoryLabel from '../../components/CategoryLabel';
import './style.css';

class PostDetails extends Component {
  constructor(props) {
    super(props);

    this.handleOpen = this.handleOpen.bind(this);
    this.handleClose = this.handleClose.bind(this);

    this.state = {
      modalOpen: false,
      visible: true,
    };
  }

  getPostContent() {
    const { post } = this.props;
    if (post.content.protected === false) {
      return post.content.rendered;
    }

    return (
      "<p>This post is password protected! Enter the password <a href='" +
      post.link +
      "?pwapp_theme_mode=desktop'>here</a>."
    );
  }

  handleOpen() {
    this.setState({ modalOpen: true, visible: !this.state.visible });
  }

  handleClose() {
    this.setState({ modalOpen: false, visible: !this.state.visible });
  }

  render() {
    const { post } = this.props;
    const { author } = post._embedded;
    const categoriesList = post._embedded['wp:term'];
    const featuredMedia = post._embedded['wp:featuredmedia'];

    // adsense params
    const googleAds = config.get('googleAds');

    // set path routes
    let goBack = '/';
    let path = '';

    if (!_.isNil(this.props.category.categorySlug) && !_.isNil(this.props.category.categoryId)) {
      path = '/category/' + this.props.category.categorySlug + '/' + this.props.category.categoryId;
      goBack = path;
    }

    path = path + '/post/' + post.slug + '/' + post.id + '/comments/' + post.comment_status;

    return (
      <Container className="post">
        <Helmet>
          <link rel="canonical" href={post.link} />
        </Helmet>
        <Link to={goBack}>
          <Icon circular size="large" name="chevron left" className={featuredMedia ? 'absolute' : ''} />
        </Link>
        {_.isArray(featuredMedia) && !_.isNil(featuredMedia[0]) ? (
          <div
            className="post-image"
            style={{ backgroundImage: `url(${featuredMedia[0].source_url})` }}
          />
        ) : null}
        <Container textAlign="justified">
          <Header>
            <h1 dangerouslySetInnerHTML={{ __html: post.title.rendered }} />
          </Header>
          {categoriesList[0].map(category => (
            <CategoryLabel
              link={`/category/${category.slug}/${category.id}`}
              name={category.name}
              key={category.name}
            />
          ))}
          <Header.Subheader>
            &nbsp;{this.props.texts.TEXTS && this.props.texts.TEXTS.BY_AUTHOR}&nbsp;<b>
              {author.map(item => item.name).join(', ')}
            </b>,&nbsp;<Moment format="MMMM DD, YYYY">{post.date}</Moment>
          </Header.Subheader>
          <div
            className="post-content"
            dangerouslySetInnerHTML={{ __html: this.getPostContent() }}
          />
          {!_.isNil(googleAds) && !_.isNil(googleAds.phone) ? (
            <DFPSlotsProvider
              dfpNetworkId={googleAds.phone.networkCode}
              sizeMapping={[{ viewport: [900, 768], sizes: googleAds.phone.sizes }]}
            >
              <div className="mobile-ads">
                <AdSlot sizes={googleAds.phone.sizes} adUnit={googleAds.phone.adUnitCode} />
              </div>
            </DFPSlotsProvider>
          ) : null}
        </Container>
        {post.content.protected === false ? (
          <Modal
            className="share"
            open={this.state.modalOpen}
            onClick={this.handleClose}
            trigger={
              <Transition.Group animation="fade up" duration="500" className="transition-container">
                {this.state.visible && (
                  <Icon circular name="share alternate" size="large" onClick={this.handleOpen} />
                )}
              </Transition.Group>
            }
            basic
          >
            <Modal.Actions>
              <Link to={path}>
                <Icon circular name="comment" size="large" />
              </Link>
              <SocialMedia title={post.title.rendered} link={post.link} />
            </Modal.Actions>
          </Modal>
        ) : null}
      </Container>
    );
  }
}

PostDetails.propTypes = {
  post: postPropType.isRequired,
  category: PropTypes.shape({
    categorySlug: PropTypes.string,
    categoryId: PropTypes.string,
  }).isRequired,
  texts: PropTypes.shape({
    TEXTS: PropTypes.shape({
      BY_AUTHOR: PropTypes.string,
    }),
  }),
};

PostDetails.defaultProps = {
  texts: {
    TEXTS: {
      BY_AUTHOR: 'by',
    },
  },
};

export default PostDetails;
