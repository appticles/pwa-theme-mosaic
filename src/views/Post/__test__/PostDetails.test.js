import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import PostDetails from '../PostDetails';

jest.mock('../components/SocialMedia', () => 'Social media test');
jest.mock('react-global-configuration', () => require.requireActual('../../../__mocks__/react-global-configuration').default);

describe('Post details', () => {
  it('should render post without image', () => {
    const mockPost = {
      id: 1,
      title: {
        rendered: 'Mock title',
      },
      author: 45,
      _embedded: {
        author: [{ id: 45, name: 'Some mock author' }],
        'wp:term': [[
          { id: 1, name: 'Some category', slug: 'some-category-slug-1' },
          { id: 2, name: 'Some category', slug: 'some-category-slug-2' },
          { id: 32, name: 'Some category', slug: 'some-category-slug-32' },
        ]],
      },
      categories: [2, 54, 23, 6],
      slug: 'mock-slug',
      content: {
        rendered: 'Some content that should be rendered',
        protected: false,
      },
      excerpt: {
        rendered: 'Some excerpt to be rendered',
        protected: true,
      },
      link: 'some-post-link',
      date: 'someMockDate',
    };
    const category = {
      match: {
        params: {
          categoryId: '32',
          categorySlug: 'some-slug',
        },
      },
    };
    const output = shallow(<PostDetails post={mockPost} category={category} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should render post with featured image', () => {
    const mockPost = {
      id: 1,
      title: {
        rendered: 'Mock title',
      },
      author: 45,
      featured_media: 23,
      _embedded: {
        author: [{ id: 45, name: 'Some mock author' }],
        'wp:term': [[
          { id: 1, name: 'Some category', slug: 'some-category-slug-1' },
          { id: 2, name: 'Some category', slug: 'some-category-slug-2' },
          { id: 32, name: 'Some category', slug: 'some-category-slug-32' },
        ]],
        'wp:featuredmedia': [
          {
            id: 23,
            source_url: 'some-post-image',
          },
        ],
      },
      categories: [2, 54, 23, 6],
      slug: 'mock-slug',
      content: {
        rendered: 'Some content that should be rendered',
        protected: false,
      },
      excerpt: {
        rendered: 'Some excerpt to be rendered',
        protected: true,
      },
      link: 'some-post-link',
      date: 'someMockDate',
    };

    const category = {
      match: {
        params: {
          categoryId: '32',
          categorySlug: 'some-slug',
        },
      },
    };
    const output = shallow(<PostDetails post={mockPost} category={category} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});
