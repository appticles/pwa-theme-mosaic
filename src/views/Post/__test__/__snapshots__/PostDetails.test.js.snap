// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post details should render post with featured image 1`] = `
<Container
  className="post"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <link
      href="some-post-link"
      rel="canonical"
    />
  </HelmetWrapper>
  <Link
    replace={false}
    to="/"
  >
    <Icon
      as="i"
      circular={true}
      className="absolute"
      name="chevron left"
      size="large"
    />
  </Link>
  <div
    className="post-image"
    style={
      Object {
        "backgroundImage": "url(some-post-image)",
      }
    }
  />
  <Container
    textAlign="justified"
  >
    <Header>
      <h1
        dangerouslySetInnerHTML={
          Object {
            "__html": "Mock title",
          }
        }
      />
    </Header>
    <Link
      key="Some category"
      replace={false}
      to="/category/some-category-slug-1/1"
    >
      <Label>
        Some category
      </Label>
    </Link>
    <Link
      key="Some category"
      replace={false}
      to="/category/some-category-slug-2/2"
    >
      <Label>
        Some category
      </Label>
    </Link>
    <Link
      key="Some category"
      replace={false}
      to="/category/some-category-slug-32/32"
    >
      <Label>
        Some category
      </Label>
    </Link>
    <HeaderSubheader>
       
      by
       
      <b>
        Some mock author
      </b>
      , 
      <t
        ago={false}
        calendar={false}
        element="time"
        filter={[Function]}
        format="MMMM DD, YYYY"
        fromNow={false}
        interval={60000}
        onChange={[Function]}
        toNow={false}
        unix={false}
        utc={false}
      >
        someMockDate
      </t>
    </HeaderSubheader>
    <div
      className="post-content"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Some content that should be rendered",
        }
      }
    />
  </Container>
  <Modal
    basic={true}
    className="share"
    closeOnDimmerClick={true}
    closeOnDocumentClick={false}
    dimmer={true}
    onClick={[Function]}
    open={false}
    trigger={
      <TransitionGroup
        animation="fade up"
        className="transition-container"
        duration="500"
      >
        <Icon
          as="i"
          circular={true}
          name="share alternate"
          onClick={[Function]}
          size="large"
        />
      </TransitionGroup>
    }
  >
    <ModalActions>
      <Link
        replace={false}
        to="/post/mock-slug/1/comments/undefined"
      >
        <Icon
          as="i"
          circular={true}
          color="grey"
          inverted={true}
          name="comment"
          size="large"
        />
      </Link>
      <Social media test
        link="some-post-link"
        title="Mock title"
      />
    </ModalActions>
  </Modal>
</Container>
`;

exports[`Post details should render post without image 1`] = `
<Container
  className="post"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <link
      href="some-post-link"
      rel="canonical"
    />
  </HelmetWrapper>
  <Link
    replace={false}
    to="/"
  >
    <Icon
      as="i"
      circular={true}
      className=""
      name="chevron left"
      size="large"
    />
  </Link>
  <Container
    textAlign="justified"
  >
    <Header>
      <h1
        dangerouslySetInnerHTML={
          Object {
            "__html": "Mock title",
          }
        }
      />
    </Header>
    <Link
      key="Some category"
      replace={false}
      to="/category/some-category-slug-1/1"
    >
      <Label>
        Some category
      </Label>
    </Link>
    <Link
      key="Some category"
      replace={false}
      to="/category/some-category-slug-2/2"
    >
      <Label>
        Some category
      </Label>
    </Link>
    <Link
      key="Some category"
      replace={false}
      to="/category/some-category-slug-32/32"
    >
      <Label>
        Some category
      </Label>
    </Link>
    <HeaderSubheader>
       
      by
       
      <b>
        Some mock author
      </b>
      , 
      <t
        ago={false}
        calendar={false}
        element="time"
        filter={[Function]}
        format="MMMM DD, YYYY"
        fromNow={false}
        interval={60000}
        onChange={[Function]}
        toNow={false}
        unix={false}
        utc={false}
      >
        someMockDate
      </t>
    </HeaderSubheader>
    <div
      className="post-content"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Some content that should be rendered",
        }
      }
    />
  </Container>
  <Modal
    basic={true}
    className="share"
    closeOnDimmerClick={true}
    closeOnDocumentClick={false}
    dimmer={true}
    onClick={[Function]}
    open={false}
    trigger={
      <TransitionGroup
        animation="fade up"
        className="transition-container"
        duration="500"
      >
        <Icon
          as="i"
          circular={true}
          name="share alternate"
          onClick={[Function]}
          size="large"
        />
      </TransitionGroup>
    }
  >
    <ModalActions>
      <Link
        replace={false}
        to="/post/mock-slug/1/comments/undefined"
      >
        <Icon
          as="i"
          circular={true}
          color="grey"
          inverted={true}
          name="comment"
          size="large"
        />
      </Link>
      <Social media test
        link="some-post-link"
        title="Mock title"
      />
    </ModalActions>
  </Modal>
</Container>
`;
